### PR TITLE
Bug fixes notes and text note editor

### LINKED ISSUE #61: Bug fixes and clean-ups related to notes and especially text note editor
There are some bugs related to notes handling and especially the text note editor:

**TextNoteEditor:**

- Text lines overlap when a text format is applied and new lines are inserted.
- Currently empty notes are saved sometimes (only happening whenever an already existing note is editted by clicking on its view in the RecyclerListView (in NotesFragment).
- The underline text format is buggy somehow right now: the underline-span is not correctly applied and unapplied.

**NoteDAO:**

-  Current time values for createDate and modDate assignments are wrong (it is currently showing year 1970 e.g.).

**Clean-up:**
- Data and UI are not cleanly separated right now in note-related classes: a NoteModel is needed, where all data-handling will be done.
- Note-Object params (createDate and modDate) should not be declared manually but automatically assigned whenever creating a new note-row in the database.

### PR DESCRIPTION
**Pull request by xSabrina**
### Description

Closes #61
And also affects #60 (solving it for NotesFragment-list at least)

Some notes related bugs have been fixed and a feature to select and delete multiple items of the RecyclerListView in NotesFragment has been added.

### Affects

Since some things are related to database (NoteModel has been created etc.) and the list of notes connected to a book is a similar case (and therefore my changes might be important for that, too) @SarahKurek is assigned.

### Notes for Reviewer

Here are some explainations to my changes:

**_TextNoteEditor:_**

- **Text lines overlapped** when a text format was applied sometimes:  This was **caused by the way the span start position** was declared in the onTextChanged. It is now declared as **one index number lower than the current cursor position** (so used formatting options are basically applied onTextChange on the last character that has been inserted). If the last user input is a deleting a charcter instead of inserting (checked by text length before and after input) the format is not applied of course.

- **Empty notes were saved** when an already existing note has been editted: This was **caused by the way the text was checked** for being empty: the html string containing tags (linebreaks etc.) was checked and this has been changed to **now only checking the raw text using jsoup**.

- The **underline text format** was not correctly applied and unapplied. **Wrong continue statements** in quoteValid and quoteInvalid are now **return statements**.

- Adjusted **apply-cases** for the toolbar features for the RichTextEditor in TextNoteEditorFragment: instead of using a default boolean to declare selected and unselected items, which was causing bugs in complicated cases, the **toolbar icon views are now declared selected** or not (using view-method **isSelected()**) and their booleans are passed onClick to RichTextEditor. Only **undo and redo are exceptions** and only get backgroundcolor-changes on click and automatically switch back after a short time.

- Fix the **quote span appearance** (color, stripewidth, gapwidth between stripe and first character) and small bugs on its appliance.

- Fix a bug where **applied text formats did get applied to the entire rest of the text**, which was causing bugs whenver text format has been changed and new text has been added somewhere in between the existing text. The **endOfString-variable** in onTextChanged is **adjusted** now.

**_NoteDAO:_**

- Current **time values for createDate and modDate** assignments **are wrong**  (always showed any date in 1970): usage of  "System.currentTimeMillis() / 1_000L" is replaced by **Date.getTime()**; .

**_Clean-up:_**

- **Data and UI were not cleanly separated** in note-related classes: a **NoteModel with necessary methods has been created** and is now used for any data-handling in other note-related classes.

- Note-Object params (**createDate and modDate**) were **declared manually** and are **now automatically assigned** whenever creating a new note-row in the database.

**_Added feature:_**

- **Selecting and deleting multiple note items from recyclerListView** (in NotesFragment): On **longClick** an item is selected. If at least one item is selected a **panel with red background** is shown on the bottom (**bottom-up animation**) and hidden (by **bottom-down-animation**) whenever selected items are removed or all items are deselected. The panel contains an **ImageButton** (delete-icon as drawable) that **removes all selected items** from the database and the recyclerViewList onClick.

Please review & comment/approve.

### CODE REVIEW

File: BibBuddy/app/src/main/java/de/bibbuddy/NoteDAO.java [hunk 1]
```diff
@@ -139,19 +140,18 @@ public void delete(Long id) {
 	 * @param type       note type (text, voice, image)
 	 * @param text       note text if it is of type text
 	 * @param createDate note creation date
-	 * @param modDate    note modification date
 	 * @param noteFileId note file id
 	 */
-	public void updateNote(Long id, String name, int type, String text, Long createDate, Long modDate,
-												 Long noteFileId) {
+	public void updateNote(Long id, String name, int type, String text, Long createDate, Long noteFileId) {
+		Long currentTime = new Date().getTime();
```
Comment (SarahKurek) [inline BibBuddy/app/src/main/java/de/bibbuddy/NoteDAO.java:146]: If Note is update in the DB the mod_date is automatically updated and the create_date must only be set if the Note is created at first time.

**Review by SarahKurek — CHANGES_REQUESTED** [review]
Code contains some bugs that should be clarified in a personal meeting.

File: BibBuddy/app/src/main/java/de/bibbuddy/NoteDAO.java (hunk 1 above)
Comment (xSabrina) [reply BibBuddy/app/src/main/java/de/bibbuddy/NoteDAO.java:146]: There is no any other modDate update for me - ModDate is only assigned when a new note is created manually in NoteDAO. After that it needs to be changed to the value that matches the moment the user stops editting it (leaving TextNoteEditorFragment) -for this purpose there is only a NoteDAO.update() used, which is using database.update()-method, which needs all contentValues as params, that need to be changed (including the modDate). 
Why the createDate is assigned to the change-contentValues: I only assumed it's better to put all values for each column of the row into the contentValues that are put into the database, I think we can remove that though (but we should do that for all never changing values then (this means: Id, Type, CreateDate and NoteFileId).

**Comment by xSabrina** [comment]
> 
> 
> Code contains some bugs that should be clarified in a personal meeting.

Adjusted the code, in specific following:

- Adjust noteDAO.updateNote(): only put the content values that are changeable (name, text, modDate).
- Adjust setting of note-name to the first line (raw text without any tags) of the note-text
- Make separation of UI and data even clearer by adjusting NoteModel and NoteItem(/NoteTextItem/NoteAudioItem/NoteImageItem)
- Use custom backButtonHandling in TextNoteEditorFragment instead of onPause() to save notes on leaving the editorFragment.

File: BibBuddy/app/src/main/java/de/bibbuddy/BookNotesViewModel.java [hunk 2]
```diff
@@ -27,11 +27,11 @@ public BookNotesViewModel(Context context) {
         noteList = new ArrayList<>();
         for (Note item : list) {
             if (item.getType() == 1) { //text
-                noteList.add(new NoteItem(item.getName(), item.getId()));
+               // noteList.add(new NoteItem(item.getName(), item.getId()));
```
Comment (SarahKurek) [inline BibBuddy/app/src/main/java/de/bibbuddy/BookNotesViewModel.java:30]: Why is this commented out?

Comment (xSabrina) [reply BibBuddy/app/src/main/java/de/bibbuddy/BookNotesViewModel.java:30]: This was causing bugs and entire class needed to be adjusted to noteModel-usage. This should be done now.

File: BibBuddy/app/src/main/java/de/bibbuddy/NoteDAO.java [hunk 3]
```diff
@@ -4,177 +4,212 @@
...
+
+    // Link note with book
+    public boolean linkNoteWithBook(Long bookId, Long noteId) {
+
+		try (SQLiteDatabase db = dbHelper.getWritableDatabase()) {
+			ContentValues contentValues = new ContentValues();
+			contentValues.put(DatabaseHelper.BOOK_ID, bookId);
+			contentValues.put(DatabaseHelper.NOTE_ID, noteId);
+
+			long id = db.insert(DatabaseHelper.TABLE_NAME_BOOK_NOTE_LNK, null, contentValues);
```
Comment (SarahKurek) [inline BibBuddy/app/src/main/java/de/bibbuddy/NoteDAO.java]:
```suggestion
			db.insert(DatabaseHelper.TABLE_NAME_BOOK_NOTE_LNK, null, contentValues);
```

File: BibBuddy/app/src/main/java/de/bibbuddy/NoteDAO.java [hunk 4]
```diff
@@ -4,177 +4,212 @@
...
+        values.put("name", name);
+        //values.put("type", type);
+        values.put("text", text);
+        //values.put("creation_date", createDate);
+        values.put("modifikation_date", currentTime);
+        //values.put("note_file_id", noteFileId);
+        dbHelper.getWritableDatabase().update(DatabaseHelper.TABLE_NAME_NOTE, values,
+                DatabaseHelper._ID + " = ?", new String[]{String.valueOf(id)});
+        db.close();
+    }
```
Comment (SarahKurek) [inline BibBuddy/app/src/main/java/de/bibbuddy/NoteDAO.java]:
```suggestion
    public void updateNote(Long id, String name, String text) {
        Long currentTime = new Date().getTime();
        SQLiteDatabase db = dbHelper.getWritableDatabase();
        try{
           ContentValues values = new ContentValues();
           values.put("name", name);
           values.put("text", text);
           values.put("modifikation_date", currentTime);
       
        dbHelper.getWritableDatabase().update(DatabaseHelper.TABLE_NAME_NOTE, values,
                DatabaseHelper._ID + " = ?", new String[]{String.valueOf(id)});
        } finally {
          db.close();
       }
    }
```

Comment (xSabrina) [reply BibBuddy/app/src/main/java/de/bibbuddy/NoteDAO.java]: Already done - was not committed yet.

File: BibBuddy/app/src/main/java/de/bibbuddy/NoteDAO.java (hunk 3 above)
Comment (xSabrina) [reply BibBuddy/app/src/main/java/de/bibbuddy/NoteDAO.java]: Done.

File: BibBuddy/app/src/main/java/de/bibbuddy/TextNoteEditorFragment.java [hunk 5]
```diff
@@ -22,6 +30,43 @@
...
+      e.printStackTrace();
+    }
+    if (rawText.length() != 0) {
+      if(getArguments() != null) {
+        noteModel.updateNote(note, name, text);
+      } else {
+        noteModel.addNote(name, 0 , text);
+      }
+    }
+  }
```
Comment (SarahKurek) [inline BibBuddy/app/src/main/java/de/bibbuddy/TextNoteEditorFragment.java]: BookId should also be saved, which you get via a bundle.

File: BibBuddy/app/src/main/java/de/bibbuddy/TextNoteEditorFragment.java [hunk 6]
```diff
@@ -22,6 +30,43 @@
...
   private NoteModel noteModel;
 
+  @Override
+  public void onCreate(Bundle savedInstanceState)
+  {
+    super.onCreate (savedInstanceState);
+    requireActivity().getOnBackPressedDispatcher().addCallback(new OnBackPressedCallback(true) {
+      @Override
+      public void handleOnBackPressed() {
+        saveNote();
```
Comment (SarahKurek) [inline BibBuddy/app/src/main/java/de/bibbuddy/TextNoteEditorFragment.java:35]: The method for updating the note list for a book could be called here.

File: BibBuddy/.idea/codeStyles/codeStyleConfig.xml [hunk 7]
```diff
@@ -1,5 +1,5 @@
 <component name="ProjectCodeStyleConfiguration">
   <state>
-    <option name="PREFERRED_PROJECT_CODE_STYLE" value="Default" />
+    <option name="PREFERRED_PROJECT_CODE_STYLE" value="Default copy1" />
   </state>
-</component>
+</component>
```
Comment (SarahKurek) [inline BibBuddy/.idea/codeStyles/codeStyleConfig.xml:5]:
```suggestion
</component>

```

File: BibBuddy/app/src/main/java/de/bibbuddy/TextNoteEditorFragment.java (hunk 6 above)
Comment (xSabrina) [reply BibBuddy/app/src/main/java/de/bibbuddy/TextNoteEditorFragment.java:35]: On going back to BookNotesView the list is always fetched new from the database in onCreateView - there is no need to update it here, too.

File: BibBuddy/app/src/main/java/de/bibbuddy/TextNoteEditorFragment.java (hunk 5 above)
Comment (xSabrina) [reply BibBuddy/app/src/main/java/de/bibbuddy/TextNoteEditorFragment.java]: This needs multiple adjustments to other classes (especially bookView related ones) - but it is done now.

File: BibBuddy/app/src/main/res/anim/bottom_up.xml [hunk 8]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<set xmlns:android="http://schemas.android.com/apk/res/android">
+    <translate
+        android:duration="500"
+        android:fillAfter="true"
+        android:fromYDelta="75%p"
+        android:toYDelta="0%p" />
+</set>
```
Comment (SarahKurek) [inline BibBuddy/app/src/main/res/anim/bottom_up.xml:8]:
```suggestion
</set>

```

**Review by SarahKurek — APPROVED** [review]
Fits so far.

**Pull request merged**